src/Crypto/RNCryptor/V3/Decrypt.hs

```diff
@@ -116,12 +128,13 @@ decrypt :: ByteString -> ByteString -> ByteString
 decrypt input pwd =
   let (rawHdr, rest) = B.splitAt 34 input
       -- remove the hmac at the end of the file
-      (toDecrypt, _) = B.splitAt (B.length rest - 32) rest
-      hdr = parseHeader rawHdr
+      (cipherText, msgHMAC) = B.splitAt (B.length rest - 32) rest
+      hdr = parseHeader $ rawHdr
       ctx = newRNCryptorContext pwd hdr
-      clearText = decryptCBC (ctxCipher ctx) (aesIV_ . rncIV . ctxHeader $ ctx) toDecrypt
-  in  removePaddingSymbols clearText
-
+      clearText = decrypt_ (ctxCipher ctx) (rncIV . ctxHeader $ ctx) cipherText
+      hmac = makeHMAC (rncHMACSalt . ctxHeader $ ctx) pwd $ rawHdr <> cipherText
+  in
+    if consistentTimeEqual msgHMAC hmac then removePaddingSymbols clearText else error "failed decrypt, invalid HMAC"
```

Comment:
Just to note about what "consistentTimeEqual" is all about: in cryptography, when you compare things it is important that you always take exactly the same amount of time to do so. If I compare "abc" to "xxx" using standard tools, then I would compare "a" to "x" and immediately return. A clever attacker would then keep trying new strings until she found "axx" took every-so-slightly longer to return (because it tests "a == a" and then "b == y" and then returns). So now she knows the first byte, and guesses "aax" then "abx" and then it takes slightly longer and she has two bytes. She can use this to "dial in" the solution dramatically faster. If every comparison takes exactly the same amount of time (or at least the amount of time is not proportional to the "correctness" in any way), then a brute force attack is much slower since she gets no information at all in each attempt.